src/languageservice/services/validation/unused-anchors.ts

```diff
@@ -0,0 +1,82 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Red Hat. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { TextDocument } from 'vscode-languageserver-textdocument';
+import { Diagnostic, DiagnosticSeverity, DiagnosticTag, Range } from 'vscode-languageserver-types';
+import { isAlias, isCollection, isNode, isScalar, Node, Scalar, visit, YAMLMap, YAMLSeq, CST, Pair } from 'yaml';
+import { YamlNode } from '../../jsonASTTypes';
+import { SingleYAMLDocument } from '../../parser/yaml-documents';
+import { AdditionalValidator } from './types';
+import { isCollectionItem } from '../../../languageservice/utils/astUtils';
+
+export class UnusedAnchorsValidator implements AdditionalValidator {
+  validate(document: TextDocument, yamlDoc: SingleYAMLDocument): Diagnostic[] {
+    const result = [];
+    const anchors = new Set<Scalar | YAMLMap | YAMLSeq>();
+    const usedAnchors = new Set<Node>();
+    const anchorParent = new Map<Scalar | YAMLMap | YAMLSeq, Node | Pair>();
+
+    visit(yamlDoc.internalDocument, (key, node, path) => {
+      if (!isNode(node)) {
+        return;
+      }
+      if ((isCollection(node) || isScalar(node)) && node.anchor) {
+        anchors.add(node);
+        anchorParent.set(node, path[path.length - 1] as Node);
+      }
+      if (isAlias(node)) {
+        usedAnchors.add(node.resolve(yamlDoc.internalDocument));
+      }
+    });
+
+    for (const anchor of anchors) {
+      if (!usedAnchors.has(anchor)) {
+        const aToken = this.getAnchorNode(anchorParent.get(anchor));
+        if (aToken) {
+          const range = Range.create(
+            document.positionAt(aToken.offset),
+            document.positionAt(aToken.offset + aToken.source.length)
+          );
+          const warningDiagnostic = Diagnostic.create(range, `Unused anchor "${aToken.source}"`, DiagnosticSeverity.Hint, 0);
+          warningDiagnostic.tags = [DiagnosticTag.Unnecessary];
+          warningDiagnostic.data = { range, name: aToken.source };
```

Comment:
Why are we repeating the range a second time? Wouldn't `Diagnostic.range` be enough?